wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Transactions.hs

```diff
@@ -0,0 +1,162 @@
+
+import qualified Cardano.Wallet.Kernel as Kernel
+import qualified Cardano.Wallet.Kernel.DB.HdWallet as HD
+import           Cardano.Wallet.Kernel.DB.InDb (InDb (..))
+import           Cardano.Wallet.Kernel.DB.TxMeta (TxMeta (..))
+import qualified Cardano.Wallet.Kernel.DB.TxMeta as TxMeta
+import           Cardano.Wallet.WalletLayer.Types (GetTxError (..))
+
+import           Cardano.Wallet.API.Indices
+import           Cardano.Wallet.API.Request
+import qualified Cardano.Wallet.API.Request.Filter as F
+import           Cardano.Wallet.API.Request.Pagination
+import qualified Cardano.Wallet.API.Request.Sort as S
+import           Cardano.Wallet.API.Response
+import           Cardano.Wallet.API.V1.Types (V1 (..), unV1)
+import qualified Cardano.Wallet.API.V1.Types as V1
+import           GHC.TypeLits (symbolVal)
+import           Pos.Core as Core
+
+getTransactions :: MonadIO m
+                => Kernel.PassiveWallet
+                -> Maybe V1.WalletId
+                -> Maybe V1.AccountIndex
+                -> Maybe (V1 Address)
+                -> RequestParams
+                -> FilterOperations V1.Transaction
+                -> SortOperations V1.Transaction
+                -> m (Either GetTxError (WalletResponse [V1.Transaction]))
+getTransactions wallet mbWalletId mbAccountIndex mbAddress params fop sop = liftIO $ do
+    let PaginationParams{..}  = rpPaginationParams params
+    let PerPage pp = ppPerPage
+    let Page cp = ppPage
+    case castAccountFiltering mbWalletId mbAccountIndex of
+        Left err          -> return $ Left err
+        Right accountFops -> do
+            case castSorting sop of
+                Left err -> return $ Left err
+                Right mbSorting -> do
+                    db <- Kernel.getWalletSnapshot wallet
+                    -- check Pos.Core.ProtocolConstants, Core.blkSecurityParam
+                    let k = (2000 :: Word)           -- TODO: retrieve this constant from Core
+                    let currentSlot = (1000 :: Word) -- TODO: retrieve this constant from Core ??
+                    (meta, mbTotalEntries) <- TxMeta.getTxMetas
+                        (wallet ^. Kernel.walletMeta)
+                        (TxMeta.Offset . fromIntegral $ (cp - 1) * pp)
+                        (TxMeta.Limit . fromIntegral $ pp)
+                        accountFops
+                        (unV1 <$> mbAddress)
+                        (castFiltering $ mapIx unV1 <$> F.findMatchingFilterOp fop)
+                        (castFiltering $ mapIx unV1 <$> F.findMatchingFilterOp fop)
+                        mbSorting
+                    let txs = map (metaToTx db k currentSlot) meta
+                    return $ Right $ respond params txs mbTotalEntries
+
+
+-- | Type Casting for Account filtering from V1 to MetaData Types.
+castAccountFiltering :: Maybe V1.WalletId -> Maybe V1.AccountIndex -> Either GetTxError TxMeta.AccountFops
+castAccountFiltering mbWalletId mbAccountIndex =
```

Comment:
If you do the refactoring I suggested above, which is effectively using `ExceptT`, then can use `ExceptT` here also, just polymorphic in the underlying monad. That way it'll fit neatly but you still mark it as pure.